From 24789a7a9bc170784e97cbcab0fadfd1f44eb908 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Tue, 21 Jan 2025 18:49:37 +0100
Subject: [PATCH] hwrng: stm32 - improve clock management with low power
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Rework clock management to be able to go to lower power without having
initialized the RNG driver. Use a Boolean to register if the driver is
initialized or not to skip low power runtime sequences if it is no the
case.

Change-Id: I943ce82363f378be4d92e5b5492b474d19211970
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/430519
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/char/hw_random/stm32-rng.c | 41 +++++++++++++++++++++++++-----
 1 file changed, 35 insertions(+), 6 deletions(-)

diff --git a/drivers/char/hw_random/stm32-rng.c b/drivers/char/hw_random/stm32-rng.c
index 0cb084b35062..9e470f6b733e 100644
--- a/drivers/char/hw_random/stm32-rng.c
+++ b/drivers/char/hw_random/stm32-rng.c
@@ -82,6 +82,7 @@ struct stm32_rng_private {
 	const struct stm32_rng_data *data;
 	bool ced;
 	bool lock_conf;
+	bool init_done;
 };
 
 /*
@@ -293,8 +294,10 @@ static int stm32_rng_init(struct hwrng *rng)
 
 	if (priv->bus_clk) {
 		err = clk_prepare_enable(priv->bus_clk);
-		if (err)
+		if (err) {
+			clk_disable_unprepare(priv->clk);
 			return err;
+		}
 	}
 
 	/* clear error indicators */
@@ -361,21 +364,30 @@ static int stm32_rng_init(struct hwrng *rng)
 						reg & RNG_SR_DRDY,
 						10, 100000);
 
+	/* Clocks will be enabled at runtime */
+	if (priv->bus_clk)
+		clk_disable_unprepare(priv->bus_clk);
+	clk_disable_unprepare(priv->clk);
+
 	mask = RNG_SR_ERROR_MASK;
 
 	if (err || (reg & mask)) {
-		if (priv->bus_clk)
-			clk_disable_unprepare(priv->bus_clk);
-		clk_disable_unprepare(priv->clk);
 		dev_err(priv->dev, "%s: timeout:%x SR: %x!\n", __func__, err, reg);
 		return -EINVAL;
 	}
 
-	clk_disable_unprepare(priv->clk);
+	priv->init_done = true;
 
 	return 0;
 }
 
+static void stm32_rng_cleanup(struct hwrng *rng)
+{
+	struct stm32_rng_private *priv = container_of(rng, struct stm32_rng_private, rng);
+
+	priv->init_done = false;
+}
+
 static int stm32_rng_remove(struct platform_device *ofdev)
 {
 	pm_runtime_disable(&ofdev->dev);
@@ -404,10 +416,22 @@ static int __maybe_unused stm32_rng_suspend(struct device *dev)
 	struct stm32_rng_private *priv = dev_get_drvdata(dev);
 	int err;
 
+	/* Skip routine if init is not done */
+	if (!priv->init_done)
+		return 0;
+
 	err = clk_prepare_enable(priv->clk);
 	if (err)
 		return err;
 
+	if (priv->bus_clk) {
+		err = clk_prepare_enable(priv->bus_clk);
+		if (err) {
+			clk_disable_unprepare(priv->clk);
+			return err;
+		}
+	}
+
 	if (priv->data->has_cond_reset) {
 		priv->pm_conf.nscr = readl_relaxed(priv->base + RNG_NSCR);
 		priv->pm_conf.htcr = readl_relaxed(priv->base + RNG_HTCR);
@@ -459,6 +483,10 @@ static int __maybe_unused stm32_rng_resume(struct device *dev)
 	int err;
 	u32 reg;
 
+	/* Skip routine if init is not done */
+	if (!priv->init_done)
+		return 0;
+
 	err = clk_prepare_enable(priv->clk);
 	if (err)
 		return err;
@@ -608,6 +636,7 @@ static int stm32_rng_probe(struct platform_device *ofdev)
 
 	priv->rng.name = dev_driver_string(dev);
 	priv->rng.init = stm32_rng_init;
+	priv->rng.cleanup = stm32_rng_cleanup;
 	priv->rng.read = stm32_rng_read;
 	priv->rng.quality = 900;
 
@@ -617,7 +646,7 @@ static int stm32_rng_probe(struct platform_device *ofdev)
 			return PTR_ERR(priv->clk);
 
 		priv->bus_clk = devm_clk_get(&ofdev->dev, "rng_hclk");
-		if (IS_ERR(priv->clk))
+		if (IS_ERR(priv->bus_clk))
 			return PTR_ERR(priv->bus_clk);
 	} else {
 		priv->clk = devm_clk_get(&ofdev->dev, NULL);
-- 
2.50.1

